qutebrowser/completion/completionwidget.py

```diff
@@ -145,10 +145,44 @@ def __repr__(self):
     def _resize_columns(self):
         """Resize the completion columns based on column_widths."""
         width = self.size().width()
-        pixel_widths = [(width * perc // 100) for perc in self._column_widths]
         if self.verticalScrollBar().isVisible():
-            pixel_widths[-1] -= self.style().pixelMetric(
-                QStyle.PM_ScrollBarExtent) + 5
+            width -= self.style().pixelMetric(QStyle.PM_ScrollBarExtent) + 5
+        log.completion.debug("available width = {}".format(width))
+
+        widths_max = [width * perc / 100 for perc in self._column_widths]
+        widths_hints = [self.sizeHintForColumn(i)
+                for i in range(len(self._column_widths))]
+        pixel_widths = [0] * len(self._column_widths)
+        log.completion.debug("widths_hints = {}".format(widths_hints))
+        log.completion.debug("widths_max = {}".format(widths_max))
+
+        # enumerate all columns, starting with narrow ones
+        remaining_indexes = set(range(len(self._column_widths)))
+        partial_width = width
+        diffs = [widths_hints[i] - widths_max[i]
+                for i in range(len(self._column_widths))]
```

Comment:
That would make it fit to a single line, but I think that's all with respect to readability. Assigning `columns = len(self._column_widths)` at the beginning of the method would be much better.